Additional logging per #926.

briantopping opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
4. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What this PR does / why we need it**:
This adds additional logging around the rfc2136 provider. There is a commented out block that may be a better version, it should be removed if not.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Special notes for your reviewer**:
Someone with a better idea about how the documentation is laid out should review what I've wrote to try and homogenize the style with links and stuff.

**Release note**:
<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access)
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`.
-->
```release-note
NONE
```

munnerz commented on docs/tutorials/acme/rfc2136-provider.rst:
```diff
@@ -0,0 +1,171 @@
+
+::
+
+   dnssec-keygen -r /dev/urandom -a HMAC-SHA512 -b 512 -n HOST example-com-secret
```
Is this an example of the command that needs to be run as part of step (1) (`Set up your DNS server for secure dynamic updates`)?

I'm slightly confused about the distinction between the two, and the YAML example provided at the start of step (2) also references `<key name used in `dnssec-keygen`, use something semantically meaningful in both environments>`

briantopping replied:
Thanks! The feedback is critical and I was hoping to have the peer review and edits.

This is indeed necessary for setting up the DNS server. The key that is generated here needs to be provided to the DNS server as a shared secret so the signatures can be validated.

munnerz commented on docs/tutorials/acme/rfc2136-provider.rst:
```diff
@@ -0,0 +1,171 @@
+------------
+
+This configuration so far will actually do nothing. You still have to
+request a certificate!
```
Can we link off to the ACME DNS01 tutorial/user guide here, so the user has a place to go 😄

briantopping replied:
Yes, that was the goal, I don't know RST, this code was all generated by a markdown-to-RST site, lol...

munnerz commented on pkg/issuer/acme/dns/rfc2136/rfc2136.go:
```diff
@@ -155,6 +157,12 @@ func NewDNSProviderCredentials(nameserver, tsigAlgorithm, tsigKeyName, tsigSecre
+	glog.V(5).Infof("DNSProvider nameserver:       %s\n", d.nameserver)
+	glog.V(5).Infof("            tsigAlgorithm:    %s\n", d.tsigAlgorithm)
+	glog.V(5).Infof("            tsigKeyName:      %s\n", d.tsigKeyName)
+	glog.V(5).Infof("            tsigSecret:       %s\n", d.tsigSecret)
```
Is it a good idea to print all this info? Seems like some of it is/may be sensitive?

briantopping replied:
I thought about that. It seemed that because it was at `glog.V(5)`, only an admin would have access to it. Furthermore, because of the multiple levels of `base64` encoding and the cognitive load of getting it all correct before anything works, it seemed relevant to print out the actual key so it could be compared. Of course a key fingerprint would be ideal, but I don't know any standards for that in an arbitrary key like is generated here. WDYT?

briantopping replied:
What if we put **** in the middle of the key? It's probabilistically impossible that the beginning and end of a key would match without the middle, as well that the middle of the key could be guessed from the ends. Seems to solve both problems!

munnerz commented on pkg/issuer/acme/dns/util/wait.go:
```diff
@@ -94,19 +94,22 @@ func checkDNSPropagation(fqdn, value string, nameservers []string) (bool, error)
 		if !(r.Rcode == dns.RcodeSuccess || r.Rcode == dns.RcodeNameError) {
-			return false, fmt.Errorf("NS %s returned %s for %s", ns, dns.RcodeToString[r.Rcode], fqdn)
+			errs = append(errs, fmt.Errorf("NS %s returned %s for %s", ns, dns.RcodeToString[r.Rcode], fqdn))
+			continue
```
Looks like we're changing the behaviour here to try *each* NS instead of failing when one fails?

This function is called by checkDNSPropegation above (https://github.com/jetstack/cert-manager/blob/abfbb36a48e4ddea4e00c5df87cb26061a1013cc/pkg/issuer/acme/dns/util/wait.go#L77) which has the comment: `// checkDNSPropagation checks if the expected TXT record has been propagated to all authoritative nameservers.`

It seems like your change here would violate the above functions description.

briantopping replied:
Ahhh, good catch. You are correct that I was changing the behavior because a) one of my servers as you may remember was a dead/AWOL server and the process would never complete. 

In my failure case, there wasn't enough information to determine why it was failing, so maybe if we kept the original behavior but logged more origin information, I would have caught that the dead server was in the SOA/NS RRs, fixed it and solved the issue.

Is this related to your mention on chat that all servers agreeing led to a higher probability of a remote server also agreeing? I'm poorly paraphrasing that chat here...

briantopping replied:
Cool. Intuitively, seems like we could check all the servers and summarize the responses. For the untrained eye, the summary would capture the global view rather than having to parse where the retry begins and ends.

It will take a few days to get around to that, just wanted to keep a note here.

munnerz commented on pkg/issuer/acme/dns/util/wait.go:
```diff
@@ -115,15 +118,64 @@ func checkAuthoritativeNss(fqdn, value string, nameservers []string) (bool, erro
+//	}
+//
+//	return true, nil
+//}
```
Don't think this commented function was meant to be included?

briantopping replied:
Yes, I was leaving it in for review, I thought I made it clearer in the original commit logs that this was a draft commit, apologies!

The code is related to the change in semantics of all servers / one server. It makes more sense now and I'll clean it up with the other changes.

munnerz commented:
/ok-to-test

briantopping commented:
@munnerz Thanks for your feedback, very helpful! It's a pleasure to give something back to the project. 
* Reversed the changes back to sequential "fail-fast" checking of servers and updated the docs so users could be more versed in the expected behavior. 
* Also added some additional explanation and cleaned up some grammar there. 
* Finally modified the printing routine for TSIG key to mask out the middle of the key as described
